blockchain/standalone/tx.go

```diff
@@ -37,5 +43,57 @@ func IsCoinBaseTx(tx *wire.MsgTx) bool {
 		return false
 	}
 
+	// We need to do additional testing when treasury is enabled or a
+	// TSPEND will be recognized as a coinbase transaction. We will rely on
+	// the fact that a pre-treasury coinbase has at least 2 outputs that
+	// start with an OP_RETURN and at least one OP_TGEN. We also check the
+	// last TxIn[0].SignatureScript byte for an OP_TSPEND.
+	if isTreasuryEnabled {
+		if len(tx.TxOut) < 2 {
+			return false
+		}
+		l := len(tx.TxIn[0].SignatureScript)
+		if l == 0 || len(tx.TxOut[0].PkScript) == 0 ||
+			len(tx.TxOut[1].PkScript) == 0 {
+			return false
+		}
+		if tx.TxIn[0].SignatureScript[l-1] == txscript.OP_TSPEND &&
+			tx.TxOut[0].PkScript[0] == txscript.OP_RETURN &&
+			tx.TxOut[1].PkScript[0] == txscript.OP_TGEN {
+			return false
+		}
+	}
+
+	return true
+}
+
+// IsTreasuryBase does a minimal check to see if a transaction is a treasury
+// base.
+func IsTreasuryBase(tx *wire.MsgTx) bool {
+	if tx.Version != wire.TxVersionTreasury {
+		return false
+	}
+
+	if len(tx.TxIn) != 1 || len(tx.TxOut) != 2 {
+		return false
+	}
+
+	if len(tx.TxOut[0].PkScript) != 1 ||
+		tx.TxOut[0].PkScript[0] != txscript.OP_TADD {
+		return false
+	}
+
+	if len(tx.TxOut[1].PkScript) != 14 ||
```

Comment:
I think we can reduce the script for the OP_RETURN to just put the height of the block in there.  There is no need for the extra randomness.  There is guaranteed to be only one per block by consensus and every block must have a unique height, so that is enough to ensure a distinct hash.  It will also save an additional 8 bytes per treasury base.

This would then become `!= 6`, and `OP_DATA_4`.

Reply:
This was painful but is fixed.